design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,599 @@
+   did not include a `resources` array.
+1. The XR and one or more resources. This would be the case if either:
+   1. The Composition included both a `resources` array and a `functions` array.
+      In this case the `resources` array is computed first then passed to
+      `functions`.
+   1. This was not a newly created XR, but instead an XR that had already
+      created its composed resources and was now being updated.
+
+In the case of `Container` functions this input and output would correspond to
+stdin and stdout. This may differ for future function implementations; for
+example a hypothetical webhook function implementation might take the input as
+an HTTP PUT request body and return the output as a response body. Crossplane
+would be responsible for reading stdout from the final function and applying its
+changes to the relevant XR and subsequent composed resources.
+
+An example function that responded by asking Crossplane to create (compose) a
+`CloudSQLInstance` would do so by returning the following `ResourceList`:
+
+```yaml
+apiVersion: fn.apiextensions.crossplane.io/v1alpha1
+kind: ResourceList
+functionConfig:
+  apiVersion: database.example.org/v1alpha1
+  kind: Config
+  metadata:
+    name: cloudsql
+  spec:
+    version: POSTGRES_9_6
+items:
+- apiVersion: database.example.org/v1alpha1
+  kind: XPostgreSQLInstance
+  metadata:
+    name: my-db
+    annotations:
+      fn.apiextensions.crossplane.io/type: "CompositeResource"
+  spec:
+    parameters:
+      storageGB: 20
+    compositionSelector:
+      matchLabels:
+        provider: gcp
+    writeConnectionSecretToRef:
+      name: db-conn
+- apiVersion: database.gcp.crossplane.io/v1beta1
+  kind: CloudSQLInstance
+  metadata:
+    name: cloudsqlpostgresql
+    annotations:
+      fn.apiextensions.crossplane.io/type: "ComposedResource"
+      fn.apiextensions.crossplane.io/name: cloudsqlinstance
+  spec:
+    forProvider:
+      databaseVersion: POSTGRES_9_6
+      region: us-central1
+      settings:
+        tier: db-custom-1-3840
+        dataDiskType: PD_SSD
+        dataDiskSizeGb: 20
+    writeConnectionSecretToRef:
+      namespace: crossplane-system
+      name: cloudsqlpostgresql-conn
+results:
+- severity: Error
+  message: "Could not render Database.postgresql.crossplane.io/v1alpha1`
+```
+
+Note the Crossplane specific `fn.apiextensions.crossplane.io/` annotations:
+
+1. `type` distinguishes an XR from a composed resource. XRs are the first
+   element in the `items` array by convention.
+1. `name` associates a composed resource entry in the `items` array with its
+   corresponding entry in the `resources` array (if any).
+
+While KRM-function-like this API is not KRM function compatible. See the
+[Alternatives Considered](#alternatives-considered) section for details on why.
+
+### Running Container Function Pipelines
+
+Crossplane is almost always deployed as a container in a Kubernetes Pod. This
```

Comment:
Going with my previous comments, Crossplane running functions or pipelines of functions shouldn't be a thing in my opinion. Crossplane doing a sequence of HTTP requests to orchestrate the functions required by a composition sounds like a much simpler and cleaner approach.

Reply:
I'm going to resolve this one since you have a handful of other comment threads around the same concern.